src/libraries/Common/src/Microsoft/Win32/SafeHandles/OverlappedValueTaskSource.Windows.cs

```diff
@@ -0,0 +1,216 @@
+            if (result != 0)
+            {
+                Complete(errorCode: (uint)result, numBytes: (uint)(result >> 32) & 0x7FFFFFFF);
+            }
+        }
+
+        /// <summary>Invoked when the asynchronous operation has completed asynchronously.</summary>
+        private static void IOCallback(uint errorCode, uint numBytes, NativeOverlapped* pOverlapped)
+        {
+            OverlappedValueTaskSource? vts = (OverlappedValueTaskSource?)ThreadPoolBoundHandle.GetNativeOverlappedState(pOverlapped);
+            Debug.Assert(vts is not null);
+            Debug.Assert(vts._overlapped == pOverlapped, "Overlaps don't match");
+
+            // Set the value to a packed combination of the error code and number of bytes (plus a high-bit 1
+            // to ensure the value we're setting is non-zero).  If it was already non-0 (the common case), then
+            // the call site already finished scheduling the async operation, in which case we're ready to complete.
+            Debug.Assert(numBytes < int.MaxValue);
+            if (Interlocked.Exchange(ref vts._result, (1ul << 63) | ((ulong)numBytes << 32) | errorCode) != 0)
+            {
+                vts.Complete(errorCode, numBytes);
+            }
+        }
+
+        internal void Complete(uint errorCode, uint numBytes)
+        {
+            ReleaseResources();
+
+            UpdateOwnerState(errorCode, numBytes);
+
+            switch (errorCode)
+            {
+                case Interop.Errors.ERROR_SUCCESS:
+                case Interop.Errors.ERROR_PIPE_CONNECTED: // special case for when the client has already connected (used only by ConnectionValueTaskSource)
+                case Interop.Errors.ERROR_BROKEN_PIPE when _isRead: // write to broken pipe should throw, read should return 0
+                case Interop.Errors.ERROR_NO_DATA when _isRead:
+                case Interop.Errors.ERROR_PIPE_NOT_CONNECTED when _isRead: // write to disconnected pipe should throw, read should return 0
+                case Interop.Errors.ERROR_HANDLE_EOF when _isRead: // read at end of file
+                case Interop.Errors.ERROR_MORE_DATA when _isRead: // the buffer was not big enough to consume all available data
+                    // Success
+                    _source.SetResult((int)numBytes);
+                    break;
+
+                case Interop.Errors.ERROR_OPERATION_ABORTED:
+                    CancellationToken ct = _cancellationRegistration.Token;
+                    _source.SetException(ct.IsCancellationRequested ? new OperationCanceledException(ct) : new OperationCanceledException());
+                    break;
+
+                case Interop.Errors.ERROR_INVALID_HANDLE:
+                    // For invalid handles, detect the error and mark our handle
+                    // as invalid to give slightly better error messages.  Also
+                    // help ensure we avoid handle recycling bugs.
+                    _fileHandle!.SetHandleAsInvalid();
```

Comment:
in case `Dispose` happened after the operation was started, only `PipeStream` was so far disposing the handle:

https://github.com/dotnet/runtime/blob/84c2ff117e9ba648bca593d87cb69d12c661f518/src/libraries/System.IO.Pipes/src/System/IO/Pipes/PipeStream.Windows.cs#L641

`RandomAccess` was doing that only when the handle was disposed before the operation was started:

https://github.com/dotnet/runtime/blob/84c2ff117e9ba648bca593d87cb69d12c661f518/src/libraries/System.Private.CoreLib/src/System/IO/RandomAccess.Windows.cs#L272